prevent corrupted spdy stream after hijacking connection

cezarsa opened the pull request:
This PR fixes corner case in spdy stream code where some bytes would never arrive at the server.

Reading directly from a hijacked connection isn't safe because some data may have already been read by the server before `Hijack` was called. To ensure all data will be received it's safer to read from the returned `bufio.Reader`. This problem seem to happen more frequently when using Go 1.8.
This is described in https://golang.org/pkg/net/http/#Hijacker:

> // The returned bufio.Reader may contain unprocessed buffered
   // data from the client.

I came across this while debugging a flaky test that used code from the `k8s.io/apimachinery/pkg/util/httpstream/spdy` package. After filling the code with debug logs and long hours running the tests in loop in the hope of catching the error I finally caught something weird.

The first word on the first spdy frame [read by the server here](https://github.com/kubernetes/kubernetes/blob/b625085230f384fa7f2ac161a5ec15cd8dcbc896/vendor/github.com/docker/spdystream/spdy/read.go#L148) had the value `0x03000100`. See, the first frame to arrive on the server was supposed to be a control frame indicating the creation of a new stream, but all control frames need the high-order bit set to 1, which was not the case here, so the saver mistakenly assumed this was a data frame and the stream would never be created. The correct value for the first word of a SYN_STREAM frame was supposed to be `0x80030001` and this lead me on the path of finding who had consumed the first 1 byte prior to the frame reader being called and finally finding the problem with the Hijack call.

I added a new test to try stressing this condition and ensuring that this bug doesn't happen anymore. However, it's quite ugly as it loops 1000 times creating streams on servers to increase the chances of this bug happening. So, I'm not sure whether it's worth it to keep this test or if I should remove it from the PR. Please let me know what you guys think and I'll be happy to update this.

Fixes #45093 #45089 #45078 #45075 #45072 #45066 #45023

cezarsa committed:
prevent corrupted spdy stream after hijacking connection

Reading directly from a hijacked connection isn't safe because some
data may have already been read by the server before Hijack was called.
To ensure all data will be received it's safer to read from the returned
bufio.Reader.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/43922)
<!-- Reviewable:end -->

ncdc commented:
@cezarsa thanks for writing this. It's been on my list but I hadn't gotten around to it yet. This was also something that confused me - are we supposed to read from the buffered reader until it's exhausted, and then switch to reading from the connection, or can we do what you did, and just always read from the buffered reader?

cezarsa commented:
@ncdc I'm glad I could help. It was an interesting one to track down. :)

From what I can see by looking at the hijack [implementation](https://github.com/golang/go/blob/go1.8/src/net/http/server.go#L298) it's safe to keep reading from the bufio.Reader, after the buffer has been exhausted it will read from the underlying connection (actually it's [backed by a `*connReader`](https://github.com/golang/go/blob/go1.8/src/net/http/server.go#L258) instance which then reads from the connection). Also, tests are passing some that's some indication that it's doing its job.

I guess there's a slight overhead of reading from the bufio.Reader because it goes through the `connReader` but in my opinion that's not enough to justify adding logic to read from the bufio.Reader and then switch the Conn directly once its been exhausted.

grodrigues3 commented:
@k8s-bot ok to test

smarterclayton commented:
/approve

cezarsa commented:
I'm not sure about what went wrong in the tests. It looks like the apiserver [didn't even start](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/43922/pull-kubernetes-e2e-gce/26395/nodelog?junit=junit_runner.xml&wrap=on) on `Jenkins GCE e2e` tests. Is this my fault? Should I rebase the code? Any pointers on how to deal with this are welcome. :)

ncdc commented:
@k8s-bot cvm gce e2e test this
@k8s-bot gce etcd3 e2e test this

ncdc commented:
> I added a new test to try stressing this condition and ensuring that this bug doesn't happen anymore. However, it's quite ugly as it loops 1000 times creating streams on servers to increase the chances of this bug happening. So, I'm not sure whether it's worth it to keep this test or if I should remove it from the PR. Please let me know what you guys think and I'll be happy to update this.

I'm a bit hesitant to include this test. How frequently does it fail without your fix?

cezarsa commented:
Sorry for the late response. The added test always fails with Go 1.8, however I couldn't trigger a failure with Go 1.7. I'm fine removing it as it was most useful when debugging this.

ncdc commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/upgrade.go:
```diff
@@ -32,6 +34,19 @@ const HeaderSpdy31 = "SPDY/3.1"
 type responseUpgrader struct {
 }
 
+// connWrapper is used to wrap a hijacked connection its bufio.Reader. All
```
Could you please put an "and" in between "connection" and "its"?

ncdc commented:
OK, thanks for the info. @liggitt do you think we should keep or remove this new test?

liggitt commented:
I'd rather have a deterministic non-stress test for the unit test. If that's not possible, I'd probably remove this one.

> From what I can see by looking at the hijack implementation it's safe to keep reading from the bufio.Reader

This is slightly concerning... is the behavior of the returned buffer documented to work that way, or are we depending on an implementation detail that can change?

ncdc commented:
> I'd rather have a deterministic non-stress test for the unit test. If that's not possible, I'd probably remove this one.

AFAIK, this is not possible.

> This is slightly concerning... is the behavior of the returned buffer documented to work that way, or are we depending on an implementation detail that can change?

Here is what is in godoc:

```go
// The returned bufio.Reader may contain unprocessed buffered
// data from the client.
Hijack() (net.Conn, *bufio.ReadWriter, error)
```

I wonder if it would be safer to read from the `bufio.ReadWriter` until it's exhausted, then switch to reading from the `net.Conn`?

Or we could ask the go team what we should do.

ncdc commented:
BTW this was guidance from the go team: https://github.com/kubernetes/kubernetes/issues/38228#issuecomment-272591961

ncdc commented:
@dsnet could you please review this PR and let us know if it's acceptable re the use of the buffered reader returned by the Hijack() call, or if we need to do something differently? Thanks!

cezarsa commented:
I'll remove the test. Regarding reading from the bufio, if we were to peek from the buffer something like this could be made maintain the wrapper idea:

```go
func (w *connWrapper) Read(b []byte) (n int, err error) {
	if w.bufReader != nil {
		w.bufData, err = w.bufReader.Peek(w.bufReader.Buffered())
		if err != nil {
			return
		}
		w.bufReader = nil
	}
	if len(w.bufData) > 0 {
		n = copy(b, w.bufData)
		w.bufData = w.bufData[n:]
		return
	}
	return w.Conn.Read(b)
}
```
It's a bit more convoluted then reading straight from the bufio but it works.

ncdc commented:
That certainly sounds better

cezarsa commented:
Okay, but since using a io.MultiReader with the bufio and the net.Conn directly would not work because the bufio would never EOF, we have to create a bytes.Reader with the peeked []byte to use in the MultiReader:

```go
func (w *connWrapper) Read(b []byte) (n int, err error) {
	if w.reader == nil {
		var bufData []byte
		bufData, err = w.bufReader.Peek(w.bufReader.Buffered())
		if err != nil {
			return
		}
		w.reader = io.MultiReader(bytes.NewReader(bufData), w.Conn)
	}
	return w.reader.Read(b)
}
```

dsnet commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/upgrade.go:
```diff
@@ -32,6 +34,19 @@ const HeaderSpdy31 = "SPDY/3.1"
+	bufReader *bufio.Reader
+}
+
+func (w *connWrapper) Read(b []byte) (n int, err error) {
```
This solution is okay, but has one edge case. When Close is called, it is entirely possible for Read to still return with data instead of EOF. That may or may not be surprising behavior.

To fix this, you could discard the buffer upon Close.
```go
func (w *connWrapper) Close() error {
    w.bufReader.Discard(w.bufReader.Buffered())
    return w.Conn.Close()
}
```

cezarsa commented:
@dsnet, thanks for the insight on handling the Close method. I have just updated the code to include your suggestion, however, since the connWrapper instance is sent as a net.Conn to an [external library](https://github.com/docker/spdystream), it's entirely possible for it to have concurrent calls to Close and Read. To fix this race I added a mutex to both methods protecting access to the bufReader instance.

ncdc commented:
We don't want to have a lock in Read(). Need to find some other solution.

cezarsa commented:
To avoid the lock, one option would be to simply remove the Close implementation and ignore the possibility of calls to Read temporarily returning data even after a Close. From what I can see this seems safe here.

Yet another option, preserving the Close semantics, would be using atomics to flag the connection as closed:

```go
func (w *connWrapper) Read(b []byte) (n int, err error) {
	if atomic.LoadInt32(&w.closed) == 1 {
		return 0, io.EOF
	}
	return w.bufReader.Read(b)
}

func (w *connWrapper) Close() error {
	err := w.Conn.Close()
	atomic.StoreInt32(&w.closed, 1)
	w.bufReader.Discard(w.bufReader.Buffered())
	return err
}
```

Even on the MultiReader approach, peeking bytes from the bufio, I can't see a way to escape from having some kind of synchronization mechanism if we were to handle Close correctly and support concurrent calls to Read and Close.

dsnet commented:
Using an atomic SGTM. In your new solution, I don't think it is necessary to call `w.bufReader.Discard` anymore.

cezarsa commented:
Updated with the atomic approach for now.

luxas commented:
Is this easier to do now that we have go1.8?
This PR should be prioritized in case it makes our implementation with Go 1.8 less race-prone

ncdc commented:
We do need to get this in soon, as the upgrade to 1.8 and/or my #44451 have made the spdy-related tests flakey.

ncdc commented:
@k8s-bot verify test this
@k8s-bot gci gce e2e test this
@k8s-bot non-cri e2e test this

ncdc commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/upgrade.go:
```diff
@@ -32,6 +36,30 @@ const HeaderSpdy31 = "SPDY/3.1"
+	if atomic.LoadInt32(&w.closed) == 1 {
+		return 0, io.EOF
+	}
+	return w.bufReader.Read(b)
```
@dsnet if we make it past the closed check on line 51, and before this line is executed, `Close()` executes and closes the `net.Conn`, what will happen when we next execute `w.bufReader.Read(b)` - will it EOF?

ncdc commented:
/release-note-none

ncdc commented:
/assign

ncdc commented:
/lgtm
